### ISSUE TITLE
faces-config.xml fails to be parsed with 3.10 and 311

### ISSUE THREAD

**Issue by artragis** [omnifaces/omnifaces#634]
Hello,

# Summary

Upgrading omnifaces from 3.9 to 3.10/3.11 we faced an issue when we start our application once deployed on any tomcat (tested on 9.0.33 and further).

# Steps to reproduce

- Use omnifaces 3.10 and Primefaces 10.0
- Use this faces-config.xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<faces-config xmlns="http://xmlns.jcp.org/xml/ns/javaee"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee http://xmlns.jcp.org/xml/ns/javaee/web-facesconfig_2_3.xsd"
	version="2.3">

	<application>
		<action-listener>org.primefaces.application.DialogActionListener</action-listener>
		<navigation-handler>org.primefaces.application.DialogNavigationHandler</navigation-handler>
		<view-handler>org.primefaces.application.DialogViewHandler</view-handler>
		<el-resolver>org.primefaces.application.exceptionhandler.PrimeExceptionHandlerELResolver</el-resolver>

		<!-- In addition with <url-pattern>/javax.faces.resource/*</url-pattern> -->
		<!-- This handler remove .xhtml extension from resources files -->
		<resource-handler>org.omnifaces.resourcehandler.UnmappedResourceHandler</resource-handler>

		<resource-bundle>
			<base-name>com.gbs.i18n.translations</base-name>
			<var>translate</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.task.translations</base-name>
			<var>translateTask</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.customization.translations</base-name>
			<var>translateCustomization</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.drilldown.translations</base-name>
			<var>translateDrilldown</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.admin_tools.translations</base-name>
			<var>translateTool</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.users.translations</base-name>
			<var>translateUser</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.servers.translations</base-name>
			<var>translateServer</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.preferences.translations</base-name>
			<var>translatePreferences</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.access_levels.translations</base-name>
			<var>translateAccessLevel</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.monitoring.translations</base-name>
			<var>translateMonitoring</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.targets.translations</base-name>
			<var>translateTarget</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.actions.translations</base-name>
			<var>translateAction</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.security.translations</base-name>
			<var>translateSecurity</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.exceptions.translations</base-name>
			<var>translateException</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.dynamic_parameters.translations</base-name>
			<var>translateDynamicParameter</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.template.translations</base-name>
			<var>translateTemplate</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.plus.translations</base-name>
			<var>translatePlus</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.univ.translations</base-name>
			<var>translateUniv</var>
		</resource-bundle>
		<resource-bundle>
			<base-name>com.gbs.i18n.bind.translations</base-name>
			<var>translateBind</var>
		</resource-bundle>
		<locale-config>
			<default-locale>en</default-locale>
			<supported-locale>en</supported-locale>
			<supported-locale>fr</supported-locale>
		</locale-config>
	</application>

	<factory>
		<exception-handler-factory>org.primefaces.application.exceptionhandler.PrimeExceptionHandlerFactory</exception-handler-factory>
	</factory>

	<lifecycle>
		<phase-listener>com.gbs.three60.suite.gui.debug.GbsPhaseListener</phase-listener>
	</lifecycle>

	<!-- Reference this custom view handler to get PrettyFaces working with 
		JSF2.3 websocket. We can remove it when PrettyFaces becomes compliant with 
		JSF2.3 -->
	<application>
		<navigation-handler>org.ocpsoft.rewrite.faces.RewriteNavigationHandler</navigation-handler>
		<view-handler>com.gbs.three60.suite.gui.jsf.url.rewriting.CustomUrlRewritingViewHandler</view-handler>
	</application>

	<!-- omnifaces compatibility issue on some Linux environments -->
	<component>
		<component-type>org.omnifaces.component.util.ResolveComponent</component-type>
		<component-class>org.omnifaces.component.util.ResolveComponent</component-class>
	</component>
	<component>
		<component-type>org.omnifaces.component.tree.Tree</component-type>
		<component-class>org.omnifaces.component.tree.Tree</component-class>
	</component>
	<component>
		<component-type>org.omnifaces.component.tree.TreeNode</component-type>
		<component-class>org.omnifaces.component.tree.TreeNode</component-class>
	</component>
	<component>
		<component-type>org.omnifaces.component.tree.TreeNodeItem</component-type>
		<component-class>org.omnifaces.component.tree.TreeNodeItem</component-class>
	</component>
	<component>
		<component-type>org.omnifaces.component.tree.TreeInsertChildren</component-type>
		<component-class>org.omnifaces.component.tree.TreeInsertChildren</component-class>
	</component>
	<converter>
		<converter-id>omnifaces.SelectItemsConverter</converter-id>
		<converter-class>org.omnifaces.converter.SelectItemsConverter</converter-class>
	</converter>
	<converter>
		<converter-id>omnifaces.SelectItemsIndexConverter</converter-id>
		<converter-class>org.omnifaces.converter.SelectItemsIndexConverter</converter-class>
	</converter>
	<converter>
		<converter-id>omnifaces.ListIndexConverter</converter-id>
		<converter-class>org.omnifaces.converter.ListIndexConverter</converter-class>
	</converter>
	<converter>
		<converter-id>omnifaces.ListConverter</converter-id>
		<converter-class>org.omnifaces.converter.ListConverter</converter-class>
	</converter>

</faces-config>

```
- Build your war and deploy on tomcat >= 9.0.33
- Restart tomcat

# Observed behaviour

- Startup fails
- The logs only say : 

```
22-Jun-2021 09:33:32.379 SEVERE [main] com.sun.faces.config.ConfigureListener.contextInitialized Critical error during deployment: 
	java.lang.IllegalStateException: OmniFaces failed to initialize! Report an issue to OmniFaces.
		at org.omnifaces.ApplicationProcessor.processEvent(ApplicationProcessor.java:73)
		at javax.faces.event.SystemEvent.processListener(SystemEvent.java:147)
		at com.sun.faces.application.applicationimpl.Events.processListeners(Events.java:277)
		at com.sun.faces.application.applicationimpl.Events.invokeListenersFor(Events.java:255)
		at com.sun.faces.application.applicationimpl.Events.publishEvent(Events.java:139)
		at com.sun.faces.application.ApplicationImpl.publishEvent(ApplicationImpl.java:151)
		at javax.faces.application.ApplicationWrapper.publishEvent(ApplicationWrapper.java:812)
		at javax.faces.application.ApplicationWrapper.publishEvent(ApplicationWrapper.java:812)
		at com.sun.faces.config.ConfigManager.publishPostConfigEvent(ConfigManager.java:577)
		at com.sun.faces.config.ConfigureListener.contextInitialized(ConfigureListener.java:288)
		at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4688)
		at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5151)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:717)
		at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:690)
		at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:705)
		at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1133)
		at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1867)
		at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
		at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
		at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:112)
		at org.apache.catalina.startup.HostConfig.deployDirectories(HostConfig.java:1045)
		at org.apache.catalina.startup.HostConfig.deployApps(HostConfig.java:429)
		at org.apache.catalina.startup.HostConfig.start(HostConfig.java:1576)
		at org.apache.catalina.startup.HostConfig.lifecycleEvent(HostConfig.java:309)
		at org.apache.catalina.util.LifecycleBase.fireLifecycleEvent(LifecycleBase.java:123)
		at org.apache.catalina.util.LifecycleBase.setStateInternal(LifecycleBase.java:423)
		at org.apache.catalina.util.LifecycleBase.setState(LifecycleBase.java:366)
		at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:936)
		at org.apache.catalina.core.StandardHost.startInternal(StandardHost.java:841)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1384)
		at org.apache.catalina.core.ContainerBase$StartChild.call(ContainerBase.java:1374)
		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
		at org.apache.tomcat.util.threads.InlineExecutorService.execute(InlineExecutorService.java:75)
		at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134)
		at org.apache.catalina.core.ContainerBase.startInternal(ContainerBase.java:909)
		at org.apache.catalina.core.StandardEngine.startInternal(StandardEngine.java:262)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		at org.apache.catalina.core.StandardService.startInternal(StandardService.java:421)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		at org.apache.catalina.core.StandardServer.startInternal(StandardServer.java:930)
		at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
		at org.apache.catalina.startup.Catalina.start(Catalina.java:633)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at org.apache.catalina.startup.Bootstrap.start(Bootstrap.java:343)
		at org.apache.catalina.startup.Bootstrap.main(Bootstrap.java:474)
	Caused by: java.lang.ExceptionInInitializerError
		at org.omnifaces.config.FacesConfigXml.instance(FacesConfigXml.java:87)
		at org.omnifaces.ApplicationProcessor.checkDuplicateResourceHandler(ApplicationProcessor.java:80)
		at org.omnifaces.ApplicationProcessor.processEvent(ApplicationProcessor.java:68)
		... 50 more
	Caused by: java.lang.IllegalStateException: FacesConfigXml failed to initialize. Perhaps your faces-config.xml contains a typo?
		at org.omnifaces.config.FacesConfigXmlSingleton.<init>(FacesConfigXmlSingleton.java:102)
		at org.omnifaces.config.FacesConfigXmlSingleton.<clinit>(FacesConfigXmlSingleton.java:58)
		... 53 more
	Caused by: org.xml.sax.SAXParseException; lineNumber: 9; columnNumber: 53; External DTD: Failed to read external DTD 'web-facesconfig_1_0.dtd', because 'http' access is not allowed due to restriction set by the accessExternalDTD property.
		at com.sun.org.apache.xerces.internal.parsers.DOMParser.parse(DOMParser.java:257)
		at com.sun.org.apache.xerces.internal.jaxp.DocumentBuilderImpl.parse(DocumentBuilderImpl.java:339)
		at javax.xml.parsers.DocumentBuilder.parse(DocumentBuilder.java:121)
		at org.omnifaces.util.Xml.parseAndAppendChildren(Xml.java:128)
		at org.omnifaces.util.Xml.createDocument(Xml.java:72)
		at org.omnifaces.config.FacesConfigXmlSingleton.loadFacesConfigXml(FacesConfigXmlSingleton.java:133)
		at org.omnifaces.config.FacesConfigXmlSingleton.<init>(FacesConfigXmlSingleton.java:95)
		... 54 more
```

# Expected behaviour

Startup succed like in 3.9

**Comment by BalusC** [comment]
Norepro.

This comment preceding a completely unnecessary block
```
<!-- omnifaces compatibility issue on some Linux environments -->
```
is a strong indicator that some libraries are messed up in your environment. Perhaps duplicate JSF or OmniFaces or even JAXP libraries elsewhere in the wrong place in the runtime classpath?

I suggest to create the smallest possible reproducer project. For example, does your issue still manifest when you remove e.g. `<var>translate</var>` bundle from `faces-config.xml`? If so, then continue removing the as many as possible pieces step by step from your project (config files, Java classes, XHTML files, dependencies, etc, etc) until your issue disappears. The last removed piece is then clearly the cause.

For reference, the sole change in `FacesConfigXml` parser between 3.9 and 3.10 is the introduction of parsing of `<resource-handler>` entries: https://github.com/omnifaces/omnifaces/commit/7c9a0dfa4cdd9bf7d9acadc9bafa1eeddea67362#diff-03e6ac08acd901ca5c3d7be0ffd8feffe3f84e0f06b46c737fd5681b065aa127 and your stack trace doesn't indicate that this method has been reached.

**Comment by artragis** [comment]
Thanks for your answer, I will try to get a minimal reproducer as soon as possible.

**Comment by melloware** [comment]
@artragis any update on this I have no issues in my apps using this.

**Comment by BalusC** [comment]
I looked at it once again, during 3.10 another change was this: https://github.com/omnifaces/omnifaces/commit/1d70b6415849be85f3678353e0a55f9301c7be20. After this, the `FacesConfigXml` parser will always be invoked during OmniFaces initialization while it was previously only lazily initialized when you used e.g. `Faces#getResourceBundles()` or `FacesConfigXml.instance().xxx()`, which you apparently never used previously.

So you had basically an existing problem in one of your faces-config.xml files which only surfaced after updating to OmniFaces 3.10. The problem is this:

> 	Caused by: org.xml.sax.SAXParseException; lineNumber: 9; columnNumber: 53; External DTD: Failed to read external DTD 'web-facesconfig_1_0.dtd', because 'http' access is not allowed due to restriction set by the accessExternalDTD property.

However, it isn't entirely clear which faces-config.xml file exactly it is referring to. Most probably it's included in one of the deployed JARs which then need updating, or it's one referenced by `javax.faces.CONFIG_FILES` context param. So I've improved the error reporting to include the exact path. It does not directly fix your problem, but it should at least give a clue how to fix it yourselves.
